chore(gateway): add gateway-broker metrics

MiguelPires opened the pull request:
## Description

Adds three metrics: latency (RTT) of gateway-broker requests, number of failed requests, total number of requests. The latency metric is visualized as a heatmap and the other two are used to calculate the failure rate of requests, which is visualized as a line graph for each request type (it can be filtered by partition and requestType).

## Related issues

closes #4487

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the release annoncement

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/impl/broker/BrokerRequestManager.java:
```diff
@@ -115,37 +116,57 @@ private static boolean responseValidation(final DirectBuffer responseContent) {
       nodeIdProvider = determineBrokerNodeIdProvider(request);
     } catch (final PartitionNotFoundException e) {
       returnFuture.completeExceptionally(e);
+      GatewayMetrics.registerFailedRequest(request.getPartitionId(), request.getType());
```
In this case, the request was never send to a broker. Do we want to include them in the metrics?

MiguelPires replied:
Why not? My reasoning was that it's still a failure on the gateway's part so it'd be useful to include it. But we can be more strict with what we consider a failure, if that allows us to extract more value from the graph. What's your opinion?

deepthidevaki replied:
That makes sense. May be, as I mentioned in another comment, we can add the reasons for failures to the metrics. So when we want to get more detailed insight, we can filter out irrelevant failures.

MiguelPires replied:
I agree, I'll add that and post some screen shots of how it looks

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/impl/broker/BrokerRequestManager.java:
```diff
@@ -115,37 +116,57 @@ private static boolean responseValidation(final DirectBuffer responseContent) {
               final BrokerResponse<T> response = request.getResponse(clientResponse);
-              handleResponse(response, returnFuture);
+
+              if (handleResponse(response, returnFuture)) {
```
Considering "RESOURCE_EXHAUSTED" as successful responses can skew the elapsedtime metrics as it is usually much smaller than a successfully processed requests.

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/impl/broker/BrokerRequestManager.java:
```diff
@@ -115,37 +116,57 @@ private static boolean responseValidation(final DirectBuffer responseContent) {
       } else if (response.isError()) {
         responseFuture.completeExceptionally(new BrokerErrorException(response.getError()));
+
+        if (response.getError().getCode() == ErrorCode.RESOURCE_EXHAUSTED) {
```
I'm not sure if this should be considered as successful.

MiguelPires replied:
Nicolas asked for it specifically. IIRC, the idea was that marking these as failed would show an increase in failures when a client sends too many requests, which isn't necessarily indicative of a problem in Zeebe. In this sense, including it would reduce the usefulness of have this metric. But @npepinpe, correct me if I'm wrong or misremembering

npepinpe replied:
The idea was that if we want to track the error rate, when stress testing RESOURCE_EXHAUSTED would show a system which is failing when it's behaving as expected.

deepthidevaki replied:
My main concern is that if there are too many resource exhausted errors during a stress test, we highly underestimate average latency because resource exhausted responses will have comparatively much lower latency.

MiguelPires replied:
We can have both by ignoring them altogether. We don't count them as failures and we also don't register their latencies.

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/metrics/GatewayMetrics.java:
```diff
@@ -0,0 +1,51 @@
+  public static void registerSuccessfulRequest(
+      final long partition, final String requestType, final long latencyMs) {
+    REQUEST_LATENCY.labels(Long.toString(partition), requestType).observe(latencyMs / 1000f);
+    TOTAL_REQUESTS.labels(Long.toString(partition), requestType).inc();
```
TOTAL_REQUESTS should be updated when the request is send, not when the response is received. Isn't it?

deepthidevaki replied:
@MiguelPires What's your opinion on this?

MiguelPires replied:
Sorry, missed this comment before. Hm, I wouldn't expect a big difference if the rate at which incoming requests come remains steady. But it seems more useful to consider TOTAL_REQUESTS as "all requests that were completed" rather than "all requests that were made including unfinished ones". If you have a lot of requests that haven't completed then I don't think we should take them into account when calculating the failure rate, since it could heavily skew the results

deepthidevaki replied:
:+1:

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/impl/broker/request/BrokerRequest.java:
```diff
@@ -63,6 +65,10 @@ public void serializeValue() {
   protected abstract T toResponseDto(DirectBuffer buffer);
 
+  public String getType() {
+    return COMMON_WORDS.matcher(this.getClass().getSimpleName()).replaceAll("");
```
Can we use the valueType and intent to build the request type instead of parsing the class name?

MiguelPires replied:
The BrokerRequest doesn't have access to the valueType and intent. This was a hack around it. It's not pretty but it works for all the request types since they all follow the same format

deepthidevaki replied:
Why not add it in `BrokerExecuteCommand`? All requests extends it, right? Anyway, it is just a suggestion.

deepthidevaki commented on gateway/src/main/java/io/zeebe/gateway/metrics/GatewayMetrics.java:
```diff
@@ -0,0 +1,51 @@
+          .labelNames("partition", "requestType")
+          .register();
+
+  private static final Counter FAILED_REQUESTS =
```
May be we can add the reasons for failure to the labels. We can then distinguish timeouts vs other failures.

deepthidevaki commented on monitor/grafana/zeebe.json:
```diff
@@ -48,8 +48,8 @@
       "type": "panel",
-      "id": "table",
-      "name": "Table",
+      "id": "table-old",
```
Is it necessary?

MiguelPires replied:
I don't know what that is but I didn't change it by hand. I exported the dashboard in the way described in the README

MiguelPires commented:
Thanks for the feedback @deepthidevaki. I refactored the logic to be clearer, added a new panel and addressed some of your comments. 

![pic](https://user-images.githubusercontent.com/3914583/87658332-54180380-c75c-11ea-92f5-75e19abf8ed5.png)

MiguelPires commented:
The failures are of the type returned by the broker, if the request fails with an exception the error type will be "unknown". I can add a special case for a TimeoutException so we can filter it

deepthidevaki commented:
> The failures are of the type returned by the broker, if the request fails with an exception the error type will be "unknown". I can add a special case for a TimeoutException so we can filter it

If it is easy to get this info, it would be good to have it. Otherwise, it is fine as it is.

MiguelPires commented:
Thanks for the suggestion @deepthidevaki, I added a case for timeouts. It'll show in grafana as "TIMEOUT" and we can filter it

MiguelPires commented:
bors r+
